src/realm/object-store/sync/sync_user.cpp

```diff
@@ -169,6 +167,67 @@ std::shared_ptr<SyncSession> SyncUser::session_for_on_disk_path(const std::strin
     return locked;
 }
 
+void SyncUser::update_state_and_tokens(SyncUser::State state, const std::string& access_token,
+                                       const std::string& refresh_token)
+{
+    std::vector<std::shared_ptr<SyncSession>> sessions_to_revive;
+    {
+        std::unique_lock<std::mutex> lock(m_mutex);
+        switch (m_state) {
+            case State::Removed:
+                REALM_ASSERT(state == State::Removed);
+                return;
+            case State::LoggedIn:
+                if (m_state == State::LoggedIn) {
```

Comment:
This is always true because the switch is already on `m_state`. Can you double check the logic in each of these cases, did you mean to switch on `state` instead?

Reply:
You're right. I've simplified this a bunch and added a unit test that makes sure these code paths are exercised.